[Order Details] Mitigate crash on shipping lines due async operations

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes WOOMOB-603

## Description
This PR mitigates the crash a merchant may experience due a race condition when they navigate through their order details but async operations that might render more rows might have not completed yet.

In some instances like shipping lines, we access the object based on `[indexPath.row]`, however we risk this to be out of bounds if some of the async operations haven't finished yet. 

## Changes

We mitigate the crash by guarding via `[safe: ]` access to the array index, following the same pattern as when accessing other rows in Order Details, like `configureCustomAmount`. With this we still log the issue in Sentry as an error, but we do not crash the app due OOB.

I've seen we also do a similar approach in different rows via `indexPath.section >= 0 && indexPath.section < sections.count`,  or `indexPath.row >= 0 && indexPath.row` (there are at least 9 matches), but we seems to use them in delegates generally.  They look pretty similar to me, so no preference of one over the other, using `[safe: ]` just seems cleaner.

From the initial report:

> Due to the complications of the order details screen, the asynchronous section reloading would cause data races that are hard to address. I'd suggest we revisit this update and find another solution that doesn't require the use of asynchronous operations here. 

Agree that the order details screen is complicated, but I think we need still need to keep async operations separated from sync ones since some rows are only rendered when we dispatch to the network. As it would require quite a big refactor through the view model, view controller, and data source for the time being we can adjust how we handle these to follow the design of the rest of the order details and just capture the errors rather than crashing on the merchant

## Testing information
- Add some async code that will update the number of shipping lines unexpectedly:
```
private func configureShippingLine(cell: HostingConfigurationTableViewCell<ShippingLineRowView>, at indexPath: IndexPath) {

  if indexPath.row == 0 && shippingLines.count > 0 {
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) { [weak self] in
      self?.order = self?.order.copy(shippingLines: []) ?? self?.order ?? Order.empty
      }
  }
... 
```
- Add a breakpoint to `ServiceLocator.crashLogging.logMessage`
- Create an order with shipping lines
- Navigate out of the order, back in, and scroll down to shipping lines
- Observe how doesn't crash, and the async call will trigger the breakpoint sending the error to sentry.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
safe array access to shipping line rows

iamgabrielma committed:
safe array access to aggregate order items line rows

iamgabrielma committed:
remove order id as unnecessary prop to log

itsmeichigo reviewed (commented):
Hi @iamgabrielma, thanks for working on this!

I don't think mitigating the crash would solve the issue. It would only stop the app from crashing but the app would show incorrect data.

In a comment in WOOMOB-603 I suggested making `reloadSections` method synchronous to avoid nasty issues. I understand the need of async check to ensure that backend receipt is available for an order. We did a similar async check for shipping label creation eligibility but in the `syncEverything` method in the view model instead. 

Do you think you can move the check there to ensure all eligibility checks are handled before reloading the table view?

iamgabrielma committed:
revert reloadsections from being async, move receipt eligibility to syncEverything

reloadSections() was async because it called await createPaymentSection(). This might create a race condition where by the time cells were configured, async operations had changed the underlying data, causing index out of bounds crashes when we attempt to render other rows that are not currently in view, like shippin lines.

iamgabrielma committed:
update VC tests to remove async usage

iamgabrielma committed:
update DS tests to remove async reloadSections calls

iamgabrielma committed:
add new tests to isEligibleForBackendReceipt exposed from DS

iamgabrielma commented:
> I don't think mitigating the crash would solve the issue. It would only stop the app from crashing but the app would show incorrect data.

Fair enough! Yeah, it wouldn't fix it, just avoid the crash and log further details in our end. 

> In a comment in [WOOMOB-603](https://linear.app/a8c/issue/WOOMOB-603) I suggested making reloadSections method synchronous to avoid nasty issues. I understand the need of async check to ensure that backend receipt is available for an order. We did a similar async check for shipping label creation eligibility but in the syncEverything method in the view model instead.
> Do you think you can move the check there to ensure all eligibility checks are handled before reloading the table view?

Definitely, and thanks for the suggestion. When originally looking into this I didn't realize we have a similar case with shipping labels. I just pushed new changes to adapt the receipt row to use this solution as well, which allows us to revert the async usage of `reloadSections()` in favor of making the receipt eligibility async check inside `syncEverything()` instead.

Recording is a bit too big for GH, so I've added them here:
- iPhone: Showing the flow to collect payment and show the receipts row async for an order that has shipping lines outside the view:  https://tilcode.blog/wp-content/uploads/2025/06/Simulator-Screen-Recording-iPhone-16-Plus-US-store-2025-06-25-at-12.54.53.mp4
- iPad: Async rendering of the receipts row from the debugger: https://tilcode.blog/wp-content/uploads/2025/06/Screen-Recording-2025-06-25-at-12.48.44.mov

I left the crash logs in the PR in case we still see them at some point 👍

itsmeichigo reviewed (approved):
Great work! Thanks for the updates 🎉

iamgabrielma committed:
Merge branch 'trunk' into fix/shipping-lines-crash-async-rows

# Conflicts:
#	WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Details/OrderDetailsDataSourceTests.swift

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/247">22.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
